File: app/src/androidTest/kotlin/com/mthaler/knittings/database/KnittingsDataSourceTest.java
package com.mthaler.knittings.database;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.filters.LargeTest;
import android.support.test.runner.AndroidJUnit4;
import com.mthaler.knittings.model.Knitting;
import com.mthaler.knittings.model.Photo;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import java.io.File;
import java.util.ArrayList;
import java.util.Date;
import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class KnittingsDataSourceTest {

    @Before
    public void initDatabase() {
        deleteAllKnittings();
    }

    @After
    public void cleanDatabase() {
        deleteAllKnittings();
    }

    @Test
    public void testAddKnitting() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        assertTrue(ds.getAllKnittings().isEmpty());
        Knitting knitting1 = ds.createKnitting("test knitting 1", "first test knitting", new Date(), null, 3.0, 42.0, 4.0);
        Knitting knitting2 = ds.createKnitting("test knitting 2", "second test knitting", new Date(), null, 4.0, 43.0, 5.0);
        Knitting knitting3 = ds.createKnitting("test knitting 3", "third test knitting", new Date(), null, 2.0, 39.0, 3.0);
        assertEquals(3, ds.getAllKnittings().size());
        ArrayList<Knitting> knittings = new ArrayList<>();
        knittings.add(knitting1);
        knittings.add(knitting2);
        knittings.add(knitting3);
        assertEquals(knittings, ds.getAllKnittings());
    }

    @Test
    public void testDeleteKnitting() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        assertTrue(ds.getAllKnittings().isEmpty());
        Knitting knitting1 = ds.createKnitting("test knitting 1", "first test knitting", new Date(), null, 3.0, 42.0, 4.0);
        Knitting knitting2 = ds.createKnitting("test knitting 2", "second test knitting", new Date(), null, 4.0, 43.0, 5.0);
        Knitting knitting3 = ds.createKnitting("test knitting 3", "third test knitting", new Date(), null, 2.0, 39.0, 3.0);
        assertEquals(3, ds.getAllKnittings().size());
        ds.deleteKnitting(knitting2);
        assertEquals(2, ds.getAllKnittings().size());
        ArrayList<Knitting> knittings = new ArrayList<>();
        knittings.add(knitting1);
        knittings.add(knitting3);
        assertEquals(knittings, ds.getAllKnittings());
    }

    @Test
    public void testUpdateKnitting() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        assertTrue(ds.getAllKnittings().isEmpty());
        Knitting knitting1 = ds.createKnitting("test knitting 1", "first test knitting", new Date(), null, 3.0, 42.0, 4.0);
        Knitting knitting2 = ds.createKnitting("test knitting 2", "second test knitting", new Date(), null, 4.0, 43.0, 5.0);
        Knitting knitting3 = ds.createKnitting("test knitting 3", "third test knitting", new Date(), null, 2.0, 39.0, 3.0);
        assertEquals(3, ds.getAllKnittings().size());
        Knitting updated = knitting2.copy(knitting2.getId(), knitting2.getTitle(), "Updated knitting", knitting2.getStarted(), knitting2.getFinished(),
                knitting2.getNeedleDiameter(), knitting2.getSize(), knitting2.getDefaultPhoto(), knitting2.getRating());
        Knitting result = ds.updateKnitting(updated);
        assertEquals(updated, result);
        assertEquals(3, ds.getAllKnittings().size());
        ArrayList<Knitting> knittings = new ArrayList<>();
        knittings.add(knitting1);
        knittings.add(updated);
        knittings.add(knitting3);
        assertEquals(knittings, ds.getAllKnittings());
    }

    @Test
    public void testGetKnitting() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        assertTrue(ds.getAllKnittings().isEmpty());
        Knitting knitting1 = ds.createKnitting("test knitting 1", "first test knitting", new Date(), null, 3.0, 42.0, 4.0);
        Knitting knitting2 = ds.createKnitting("test knitting 2", "second test knitting", new Date(), null, 4.0, 43.0, 5.0);
        Knitting knitting3 = ds.createKnitting("test knitting 3", "third test knitting", new Date(), null, 2.0, 39.0, 3.0);
        assertEquals(3, ds.getAllKnittings().size());
        assertEquals(knitting1, ds.getKnitting(knitting1.getId()));
        assertEquals(knitting2, ds.getKnitting(knitting2.getId()));
        assertEquals(knitting3, ds.getKnitting(knitting3.getId()));
    }

    @Test
    public void testCreatePhoto() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        assertTrue(ds.getAllKnittings().isEmpty());
        Knitting knitting1 = ds.createKnitting("test knitting 1", "first test knitting", new Date(), null, 3.0, 42.0, 4.0);
        Knitting knitting2 = ds.createKnitting("test knitting 2", "second test knitting", new Date(), null, 4.0, 43.0, 5.0);
        Knitting knitting3 = ds.createKnitting("test knitting 3", "third test knitting", new Date(), null, 2.0, 39.0, 3.0);
        assertEquals(3, ds.getAllKnittings().size());
        assertTrue(ds.getAllPhotos().isEmpty());
        Photo photo = ds.createPhoto(new File("/path/to/photo"), knitting2.getId(), null, "test");
        assertEquals(1, ds.getAllPhotos().size());
        ArrayList<Photo> photos = new ArrayList<>();
        photos.add(photo);
        assertEquals(photos, ds.getAllPhotos());
        assertTrue(ds.getAllPhotos(knitting1).isEmpty());
        assertEquals(1, ds.getAllPhotos(knitting2).size());
        assertEquals(photos, ds.getAllPhotos(knitting2));
        assertTrue(ds.getAllPhotos(knitting3).isEmpty());
        ds.deleteKnitting(knitting2);
        assertTrue(ds.getAllPhotos().isEmpty());
    }

    @Test
    public void testUpdatePhoto() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        assertTrue(ds.getAllKnittings().isEmpty());
        assertTrue(ds.getAllPhotos().isEmpty());
        Knitting knitting = ds.createKnitting("test knitting 1", "first test knitting", new Date(), null, 3.0, 42.0, 4.0);
        assertEquals(1, ds.getAllKnittings().size());
        assertTrue(ds.getAllPhotos().isEmpty());
        Photo photo = ds.createPhoto(new File("/path/to/photo"), knitting.getId(), null, "test");
        assertEquals(1, ds.getAllPhotos().size());
        ArrayList<Photo> photos = new ArrayList<>();
        photos.add(photo);
        assertEquals(photos, ds.getAllPhotos());
        Photo updated = photo.copy(photo.getId(), photo.getFilename(), photo.getKnittingID(), "updated description", photo.getPreview());
        ds.updatePhoto(updated);
        ArrayList<Photo> photos2 = new ArrayList<>();
        photos2.add(updated);
        assertEquals(photos2, ds.getAllPhotos());
    }

    private void deleteAllKnittings() {
        Context ctx = InstrumentationRegistry.getTargetContext();
        KnittingsDataSource ds = KnittingsDataSource.Companion.getInstance(ctx);
        for (Knitting knitting : ds.getAllKnittings()) {
            ds.deleteKnitting(knitting);
        }
    }
}

File: app/src/androidTest/java/com/mthaler/knittings/database/KnittingsDataSourceTest.kt
package com.mthaler.knittings.database

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import androidx.test.platform.app.InstrumentationRegistry
import com.mthaler.knittings.model.Knitting
import com.mthaler.knittings.model.Photo
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.File
import java.util.*
import androidx.test.ext.junit.runners.AndroidJUnit4

@RunWith(AndroidJUnit4::class)
class KnittingsDataSourceTest {

    @Before
    fun initDatabase() {
        deleteAllKnittings()
    }

    @After
    fun cleanDatabase() {
        deleteAllKnittings()
    }

    @Test
    fun testAddKnitting() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().targetContext
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3 = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        val knittings: ArrayList<Knitting> = ArrayList<Knitting>()
        knittings.add(knitting1)
        knittings.add(knitting3)
        knittings.add(knitting5)
        Assert.assertEquals(knittings, ds.allProjects)
    }

    @Test
    fun testDeleteKnitting() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3 = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        ds.deleteProject(knitting3)
        Assert.assertEquals(2, ds.allProjects.size.toLong())
    }

    @Test
    fun testUpdateKnitting() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3 = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        val updated: Knitting = knitting3.copy(knitting3.id, knitting3.title, "Updated knitting", knitting3.started, knitting3.finished, knitting3.needleDiameter,
            knitting3.size, knitting3.defaultPhoto, knitting3.rating, knitting3.duration, knitting3.category)
        val result: Knitting = ds.updateProject(updated)
        Assert.assertEquals(updated, result)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        val knittings = ArrayList<Knitting>()
        knittings.add(knitting1)
        knittings.add(updated)
        knittings.add(knitting5)
        Assert.assertEquals(knittings, ds.allProjects)
    }

    @Test
    fun testGetKnitting() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3 = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        Assert.assertEquals(knitting1, ds.getProject(knitting1.id))
        Assert.assertEquals(knitting3, ds.getProject(knitting3.id))
        Assert.assertEquals(knitting5, ds.getProject(knitting5.id))
    }

    @Test
    fun testCreatePhoto() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3 = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        Assert.assertTrue(ds.allPhotos.isEmpty())
        val photo0 = Photo(-1, File("/path/to/photo"), knitting3.id)
        val photo1: Photo = ds.addPhoto(photo0)
        Assert.assertEquals(1, ds.allPhotos.size.toLong())
        val photos = ArrayList<Photo>()
        photos.add(photo1)
        Assert.assertEquals(photos, ds.allPhotos)
        ds.deleteProject(knitting3)
        Assert.assertTrue(ds.allPhotos.isEmpty())
    }

    @Test
    fun testUpdatePhoto() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3 = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        Assert.assertTrue(ds.allPhotos.isEmpty())
        val photo0 = Photo(-1, File("/path/to/photo"), knitting3.id)
        val photo1: Photo = ds.addPhoto(photo0)
        Assert.assertEquals(1, ds.allPhotos.size.toLong())
        val photos = ArrayList<Photo>()
        photos.add(photo1)
        Assert.assertEquals(photos, ds.allPhotos)
        val updated = photo1.copy(photo1.id, photo1.filename, photo1.ownerID, "updated description", photo1.preview)
        ds.updatePhoto(updated)
        val photos2 = ArrayList<Photo>()
        photos2.add(updated)
        Assert.assertEquals(photos2, ds.allPhotos)
    }

    @Test
    fun testGetPhoto() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3: Knitting = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        Assert.assertTrue(ds.allPhotos.isEmpty())
        val photo0 = Photo(-1, File("/path/to/photo"), knitting3.id)
        val photo1: Photo = ds.addPhoto(photo0)
        Assert.assertEquals(1, ds.allPhotos.size.toLong())
        val result = ds.getPhoto(photo1.id)
        Assert.assertEquals(photo1, result)
    }

    private fun deleteAllKnittings() {
        val ctx: Context = InstrumentationRegistry.getInstrumentation().context
        val ds = KnittingsDataSource
        ds.init(ctx)
        Assert.assertTrue(ds.allProjects.isEmpty())
        val knitting0 = Knitting(-1,"test knitting 1", "first test knitting", Date(), null, "3.0", "42.0")
        val knitting1 = ds.addProject(knitting0)
        val knitting2 = Knitting(-1, "test knitting 2", "second test knitting", Date(), null, "4.0", "43.0")
        val knitting3: Knitting = ds.addProject(knitting2)
        val knitting4 = Knitting(-1, "test knitting 3", "third test knitting", Date(), null, "2.0", "39.0")
        val knitting5 = ds.addProject(knitting4)
        Assert.assertEquals(3, ds.allProjects.size.toLong())
        ds.deleteAllProjects()
        Assert.assertEquals(0, ds.allProjects.size.toLong())
        val db = ctx.databaseList()[0]
    }
}